blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md

```diff
@@ -0,0 +1,259 @@
+
+``` 
+let x = 3 in ([x: 3], \lambda env, y. let x = env.x in x + y) 
+```
+
+(here we could obviously substitute `x` in but closures are
+needed when `x` is a dynamic input).
+Can we give a type to the resulting code?  The source program had type
+`Number -> Number`, but the target had a type more like
+
+```
+{ x: Number} \times ({x : Number} \times Number -> Number).
+```
+
+In addition to being ugly, this type is leaking irrelevant details of
+the function's implementation: all of its free variables are there in
+its type, so two terms with the same function type but different free
+variables would be translated to different types.
+Also high-level program equivalences like \\(\beta\\)-reducing the term
+to just `λ y. 3 + y` would not even preserve typing.
+Not only that, but some bad code could now supply a *different* value
+for `x` than allowed which could break invariants the programmer
+had about the function.
+
+We could fix the type preservation issue by just using a dynamic type
+for our environment, but this would still leak details in the values.
+Fortunately, there is a nice solution to the other problems using more
+advanced types: existential types.
+The idea is that the type of the environment of free variables is
+*irrelevant* to anyone that calls the function, only the function
+itself should know what the environment looks like: the type of the
+environment should be *abstract* to the caller and *concrete* to the
+callee.
+Existential types capture this.
+
+We can translate functions in the source of type `A -> B` to
+pairs of an environment and a code pointer, but now making the
+environment type existentially quantified:
+
+```
+∃ Γ. Γ × (Γ × A -> B).
+```
+Then the syntax of existential types ensure that all any consumer can
+do with the `env : Γ` in the pair is pass it to the code
+pointer with an `A` argument.
+
+How do we prove that this is correct? And what does correct even mean?
+We'll focus on a property called *full abstraction* which says that if
+two programs are equal in the source language, then their translations
+are equal.
+Here, equal in the source language will just mean \\(\beta,\eta \\)
+equivalence, so things like as above:
+
+```
+let x = 3 in λ y. x + y ≡ λ y. 3 + y
+```
+
+To prove this we'll show that in a language with existential types the
+types `∃ Γ. Γ × (Γ × A -> B)` and `A \to B` are isomorphic.
+The usual proof is by parametricity, instead we'll use a closely
+related category-theoretic argument: the coYoneda lemma.
+
+## The CoYoneda Lemma
+
+The coYoneda lemma is a generalization of the equivalence described
+above.
+I'll start with the ordinary version which uses *coends* and
+*presheaves*.
+
+The coYoneda lemma says that for any category \\( C \\), *presheaf*,
+i.e. a functor \\( Q : C^{op} \to \Set \\), and object \\(A \in C \\),
+\\(Q(A) \\) is isomorphic to:
+\\[ \exists B. (A \to B) \times Q(B) \\]
+usually coends use an integral sign but I think \\(\exists \\) is a
```

Comment:
After reading the part below about "Haskelletons" and the proof going through with existentials, I realized
(1) I really don't know what a coend is
(2) I don't know how coYoneda depends on coends